src/cs/Ssh/SecureStream.cs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DevTunnels.Ssh.Events;
+
+namespace Microsoft.DevTunnels.Ssh;
+
+/// <summary>
+/// Establishes an end-to-end encrypted two-way authenticated data stream over an underlying
+/// transport stream, using the SSH protocol but providing simplified interface that is limited to
+/// a single duplex stream (channel).
+/// </summary>
+/// <remarks>
+/// This class is a complement to <see cref="MultiChannelStream"/>, which provides only the
+/// channel-multiplexing functions of SSH.
+/// 
+/// To establish a secure connection, the two sides first establish an insecure transport stream
+/// over a pipe, socket, or anything else. Then they encrypt and authenticate the connection
+/// before beginning to send and receive data.
+/// </remarks>
+public class SecureStream : Stream
+{
+	private readonly Stream transportStream;
+	private readonly SshSession session;
+	private readonly SshClientCredentials? clientCredentials;
+	private readonly SshServerCredentials? serverCredentials;
+	private SshStream? stream;
+
+	/// <summary>
+	/// Creates a secure stream over an underlying transport stream, using client credentials
+	/// to authenticate.
+	/// </summary>
+	/// <param name="transportStream">Underlying (insecure) transport stream.</param>
+	/// <param name="clientCredentials">Client authentication credentials.</param>
+	/// <param name="trace">Optional trace source for SSH protocol tracing.</param>
+	public SecureStream(
+		Stream transportStream,
+		SshClientCredentials clientCredentials,
+		TraceSource? trace = null)
+	{
+		this.transportStream = transportStream ??
```

Comment:
Let's also check that `transportStream` can read and write.

Reply:
Not necessary in my opinion, and other stream wrapper classes in this library don't do that. Streams that can't read and write are unusual, and it is sufficient to allow calls to their `Read()`/`Write()` methods to throw `NotSupportedException`.